### PR TITLE
[DO NOT SQUASH] MLIR#807: add tosa.cast to fusable elementwise ops in tosa-partition

### PR DESCRIPTION
**Pull request by sjw36**
MLIR#812: fixed bug in tosa-to-linalg-named for quantized bias add
* added test
    
    https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/812

MLIR#813: fuse all subsequent transpose/reshape ops in tosa-partition
* also make tosa-partition run bottom-up (should use a pass instead)
* changed tosa-partition=trailing-only to default true
* updated tosa-partition tests
MLIR#807: add tosa.cast to fusable elementwise ops in tosa-partition
    
    https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/807
    https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/813

* fixed rock.transform_map for empty lowerBounds
* updated test for quantization kernel in partitioning

### CODE REVIEW

File: mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td [hunk 1]
```diff
@@ -192,7 +192,7 @@ def Rock_TransformMapAttr : Rock_Attr<"TransformMap"> {
     );
 
     let assemblyFormat = [{
-        `<` $map `by` ` ` `[` $ops `]` `bounds` `=` ($upperBounds^) : (`[` `]`)? `->` $lowerBounds `>`
+        `<` $map `by` ` ` `[` $ops `]` `bounds` `=` ($upperBounds^) : (`[` `]`)? `->` ($lowerBounds^) : (`[` `]`)? `>`
```
Comment (manupak) [inline mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:199]: Why this change ?

File: external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp [hunk 2]
```diff
@@ -562,7 +567,7 @@ void TosaPartitionPass::runOnOperation() {
 
       // Grab a useful set of leading ops, like we do for trailing.
       SetVector<Operation *> leadingOps;
-      traceInputs(anchorOp, leadingOps, inputNodes);
+      traceInputs(anchorOp, anchorOp, leadingOps, inputNodes);
```
Comment (manupak) [inline external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp:569]: Why anchor op as ignore op here ?

**Review by manupak — COMMENTED** [review]
Changes broadly looks good!
just two questions..

File: mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td (hunk 1 above)
Comment (sjw36) [reply mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:199]: I found with new partitioning where an empty lower bounds happens. It's like a memref.collapse_shape : memref<1x1x1xf32> into memref<f32>

File: external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp (hunk 2 above)
Comment (sjw36) [reply external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp:569]: Since this routine is now used below for walking up inputs to trailing elementwise ops, the ignoreOp was added so we don't recurse back up the anchor op path. But here it is mostly a placeholder (vs nullptr) since we want to walk back every input of the anchor op.

File: mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td (hunk 1 above)
Comment (manupak) [reply mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:199]: Ack!

File: external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp (hunk 2 above)
Comment (manupak) [reply external/llvm-project/mlir/lib/Dialect/Tosa/Transforms/TosaPartition.cpp:569]: Ack!

**Review by manupak — APPROVED** [review]
LGTM!

**Review by pcf000 — APPROVED** [review]
A few things I did differently in my generic-partition branch;  we'll harmonise later.
